README.md
Streamio FFMPEG
===============

Simple yet powerful wrapper around the ffmpeg command for reading metadata and transcoding movies.

All work on this project is sponsored by the online video platform [Streamio](http://streamio.com).

[![Streamio](http://d253c4ja9jigvu.cloudfront.net/assets/small-logo.png)](http://streamio.com)

Installation
------------

    (sudo) gem install streamio-ffmpeg

Compatibility
-------------

### Ruby

Only guaranteed to work with MRI Ruby 1.9.3 or later.
Should work with rubinius head in 1.9 mode.
Will not work in jruby until they fix: http://goo.gl/Z4UcX (should work in the upcoming 1.7.5)

### ffmpeg

The current gem is tested against ffmpeg 1.2.1. So no guarantees with earlier (or much later) versions. Output and input standards have inconveniently changed rather a lot between versions of ffmpeg. My goal is to keep this library in sync with new versions of ffmpeg as they come along.

Usage
-----

### Require the gem

``` ruby
require 'rubygems'
require 'streamio-ffmpeg'
```

### Reading Metadata

``` ruby
movie = FFMPEG::MediaFile.new("path/to/movie.mov")

movie.duration # 7.5 (duration of the movie in seconds)
movie.bitrate # 481 (bitrate in kb/s)
movie.size # 455546 (filesize in bytes)

movie.video_stream # "h264, yuv420p, 640x480 [PAR 1:1 DAR 4:3], 371 kb/s, 16.75 fps, 15 tbr, 600 tbn, 1200 tbc" (raw video stream info)
movie.video_codec # "h264"
movie.colorspace # "yuv420p"
movie.resolution # "640x480"
movie.width # 640 (width of the movie in pixels)
movie.height # 480 (height of the movie in pixels)
movie.frame_rate # 16.72 (frames per second)

movie.audio_stream # "aac, 44100 Hz, stereo, s16, 75 kb/s" (raw audio stream info)
movie.audio_codec # "aac"
movie.audio_sample_rate # 44100
movie.audio_channels # 2

movie.valid? # true (would be false if ffmpeg fails to read the movie)
```

### Transcoding

First argument is the output file path.

``` ruby
movie.transcode("tmp/movie.mp4") # Default ffmpeg settings for mp4 format
```

Keep track of progress with an optional block.

``` ruby
movie.transcode("movie.mp4") { |progress| puts progress } # 0.2 ... 0.5 ... 1.0
```

Give custom command line options with a string.

``` ruby
movie.transcode("movie.mp4", "-ac aac -vc libx264 -ac 2 ...")
```

Use the EncodingOptions parser for humanly readable transcoding options. Below you'll find most of the supported options. Note that the :custom key will be used as is without modification so use it for any tricky business you might need.

``` ruby
options = {video_codec: "libx264", frame_rate: 10, resolution: "320x240", video_bitrate: 300, video_bitrate_tolerance: 100,
           aspect: 1.333333, keyframe_interval: 90,
           x264_vprofile: "high", x264_preset: "slow",
           audio_codec: "libfaac", audio_bitrate: 32, audio_sample_rate: 22050, audio_channels: 1,
           threads: 2,
           custom: "-vf crop=60:60:10:10"}
movie.transcode("movie.mp4", options)
```

The transcode function returns a MediaFile object for the encoded file.

``` ruby
transcoded_movie = movie.transcode("tmp/movie.flv")

transcoded_movie.video_codec # "flv"
transcoded_movie.audio_codec # "mp3"
```

Aspect ratio is added to encoding options automatically if none is specified.

``` ruby
options = { resolution: "320x180" } # Will add -aspect 1.77777777777778 to ffmpeg
```

Preserve aspect ratio on width or height by using the preserve_aspect_ratio transcoder option.

``` ruby
widescreen_movie = FFMPEG::MediaFile.new("path/to/widescreen_movie.mov")

options = { resolution: "320x240" }

transcoder_options = { preserve_aspect_ratio: :width }
widescreen_movie.transcode("movie.mp4", options, transcoder_options) # Output resolution will be 320x180

transcoder_options = { preserve_aspect_ratio: :height }
widescreen_movie.transcode("movie.mp4", options, transcoder_options) # Output resolution will be 426x240
```

For constant bitrate encoding use video_min_bitrate and video_max_bitrate with buffer_size.

``` ruby
options = {video_min_bitrate: 600, video_max_bitrate: 600, buffer_size: 2000}
movie.transcode("movie.flv", options)
```

Add watermark image on the video.

For example, you want to add a watermark on the video at right top corner with 10px padding.

``` ruby
options = { watermark: "full_path_of_watermark.png", resolution: "640x360", watermark_filter: { position: "RT", padding_x: 10, padding_y: 10 } }
```

Position can be "LT" (Left Top Corner), "RT" (Right Top Corner), "LB" (Left Bottom Corner), "RB" (Right Bottom Corner).

### Taking Screenshots

You can use the screenshot method to make taking screenshots a bit simpler.

``` ruby
movie.screenshot("screenshot.jpg")
```

The screenshot method has the very same API as transcode so the same options will work.

``` ruby
movie.screenshot("screenshot.bmp", seek_time: 5, resolution: '320x240')
```

You can preserve aspect ratio the same way as when using transcode.

``` ruby
movie.screenshot("screenshot.png", { seek_time: 2, resolution: '200x120' }, preserve_aspect_ratio: :width)
```

Specify the path to ffmpeg
--------------------------

By default, streamio assumes that the ffmpeg binary is available in the execution path and named ffmpeg and so will run commands that look something like "ffmpeg -i /path/to/input.file ...". Use the FFMPEG.ffmpeg_binary setter to specify the full path to the binary if necessary:

``` ruby
FFMPEG.ffmpeg_binary = '/usr/local/bin/ffmpeg'
```

This will cause the same command to run as "/usr/local/bin/ffmpeg -i /path/to/input.file ..." instead.


Automatically kill hung processes
---------------------------------

By default, streamio will wait for 30 seconds between IO feedback from the FFMPEG process. After which an error is logged and the process killed.
It is possible to modify this behaviour by setting a new default:

``` ruby
# Change the timeout
Transcoder.timeout = 10

# Disable the timeout altogether
Transcoder.timeout = false
```

Disabling output file validation
------------------------------

By default Transcoder validates the output file, in case you use FFMPEG for HLS
format that creates multiple outputs you can disable the validation by passing
`validate: false` to transcoder_options.

Note that transcode will not return the encoded movie object in this case since
attempting to open a (possibly) invalid output file might result in an error being raised.

```ruby
transcoder_options = { validate: false }
movie.transcode("movie.mp4", options, transcoder_options) # returns nil
```


Copyright
---------

Copyright (c) Streamio AB. See LICENSE for details.


CHANGELOG
== Master

New:
* Support watermarking (thanks smoothdvd)

Improvements:
* Allow parenthesis in colorspace (thanks walterdavis for initial code and rociiu for finding a bug with it)

== 1.0.0 2013-07-08

New:
* Bumped target ffmpeg version to 1.2.1

Improvements:
* Simpler implementation for timeouts.
  Should be far less cpu and memory dependent (don't spawn a thread for every line of output)
  Timeout spec now passes in Rubinius (using 1.9 mode)
* Give helpful error message for windows users lacking the win32-process gem (thanks casoetan)
* Add MediaFile#container (thanks vitalis)
* Support vprofile and preset encoding options (thanks vitalis)

Changes:
* Default timeout lowered to 30 seconds

Bugs:
* Avoid crash if asking for frame_rate of a video without video stream (thanks squidarth)
* Fix crash when doing audio transcoding on ffmpeg >= 1.0.1 (thanks vitalis)

Deprecations:
* Removed support for Ruby 1.8
* Removed support for ffmpeg 0.7

Refactorings:
* Quite a few, see commit history for details.

== 0.9.0 2012-07-24

New:
* Bumped target ffmpeg version to 0.11.1
* Add hung process detection with configurable timeout (thanks stakach)
* Raise FFMPEG::Error instead of generic RuntimeError on failed transcodings
* MediaFile#screenshot for more intuitive screenshotting (README has details)
* MediaFile#creation_time and MediaFile#rotation attributes when metadata is available (thanks Innonate)

Bugs:
* Fixed too many open files bug (thanks to akicho8)
* Fixed missing path escaping (thanks to mikesager)
* Fixed README typo (thanks to Linutux)
* Files outputing "could not find codec parameters" are now recognized as invalid

Deprecations:
* Removed MediaFile#uncertain_duration?
* Removed all the deprecated crop options (use :custom => '-vf crop=x:x:x:x' if you need it)

Refactorings:
* Removed the deprecated duration validation code
* Polish on the transcoder class
* Polish on the spec suite

== 0.8.5 2011-03-05

* If a clip has a DAR that doesn't make sense fall back to calculating aspect ratio from dimensions
* Allow filenames with single quote characters (thanks to youpy)

== 0.8.4 2011-11-30

* Duration now one decimal more accurate (thanks to Russel Brooks)
* Added encoding option seek_time (thanks to Misty De Meo)

== 0.8.3 2011-09-01

* Parameters now come in the order of codecs, presets, others so that we can override the presets
* Added encoding option keyframe_interval to set number of frames between i-frames (aka GOP size)
* Streamio (sponsor of this project) have launched new awesome pricing @ http://streamio.com

== 0.8.2 2011-08-19

* Path to ffmpeg binary can now be specified (thanks jonathandean)
* If ffmpeg output contains "is not supported" the MediaFile will be considered invalid

== 0.8.1 2011-07-28

* Fix progress yielding with ffmpeg 0.8
* Updated specs to pass with ffmpeg 0.8

== 0.8.0 2011-05-26

* Duration is now ALWAYS considered uncertain (we've noticed that ffmpeg is not always correct)
* This means that the duration check will normally never run (unless you manually hack @uncertain_duration to false)
* MediaFile#audio_channels now returns nil if there is no audio stream (instead of crashing)
* Development: Use Bundler
* Development: Update RSpec to 2.6

== 0.7.8 2011-04-04

* Fixed number of audio channels on files with 5.1 audio

== 0.7.7 2011-02-01

* MediaFiles with starttime are now considered as having uncertain duration as its behavior is not consistent across formats
* Upgrade development environment to RSpec 2.4

== 0.7.6 2011-01-14

* Another ruby 1.9 encoding fix

== 0.7.5 2011-01-14

* Fixed some ruby 1.9 issues
* Added MediaFile#video_bitrate and MediaFile#audio_bitrate (thanks to mbj)

== 0.7.4 2010-12-07

* Fixed broken duration on movies with start times over 0 by reducing duration with start-time

== 0.7.3 2010-08-26

* Replaced Jewler with simple dynamic gemspec file
* Spec files now not in published gem to make it a lot smaller in size
* Full output from ffmpeg command in error raised during transcoding

== 0.7.2 2010-08-11

* Added encoding option duration
* Avoid crashing when ffmpeg can't find resolution of a movie

== 0.7.1 2010-07-08

* Make sure preset parameters are always put last to avoid them ending up before any codec assignments
* Testing against a fresh ffmpeg build (r24069)

== 0.7.0 2010-07-07

* Support for ffpresets through video_preset, audio_preset and file_preset encoding options
* Added encoding option video_bitrate_tolerance

== 0.6.8.1 2010-07-06

* Bugfix - aspect ratio was not calculated properly on movies with no DAR

== 0.6.8 2010-07-06

* Don't use encoding options with nil values
* Added encoding options video_max_bitrate, video_min_bitrate and buffer_size for constant bitrate encoding

== 0.6.7 2010-06-10

* Bugfix - aspect ratio preserver could suggest non even resolutions in certain circumstances

== 0.6.6 2010-06-10

* Transcodings to .jpg and .png will now work as they will skip duration validation

== 0.6.5 2010-05-19

* MediaFile#size method to get file size.

== 0.6.4 2010-05-12

* Ruby 1.9 compatibility fix for EncodingOptions (thanks michalf!)

== 0.6.3 2010-05-05

* Use DAR to calculate aspect ratio if available

== 0.6.2 2010-05-05

* Added MediaFile#uncertain_duration? which is true if ffmpeg is guessing duration from bitrate
* Skipping the transcoders duration validation if original file has uncertain duration
* Made sure aspect ratio preservation always rounds new size to an even number to avoid "not divisible by 2" errors
* Changed MediaFile#valid? logic to accept any movie with either a readable audio or video stream

== 0.6.0 2010-05-04

* Cropping options now handled by EncodingOptions (croptop, cropbottom, cropleft and cropright)
* Aspect ratio parameter calculated and added by default
* Added transcoder options to preserve original aspect ratio on width or height

== 0.5.0 2010-04-28

* Added logging capabilities

== 0.4.3 2010-04-06


spec/ffmpeg/movie_spec.rb
require 'spec_helper.rb'

module FFMPEG
  describe MediaFile do
    describe "initializing" do
      context "given a non existing file" do
        it "should throw ArgumentError" do
          expect { MediaFile.new("i_dont_exist") }.to raise_error(Errno::ENOENT, /does not exist/)
        end
      end

      context "given a file containing a single quotation mark in the filename" do
        before(:all) do
          @movie = MediaFile.new("#{fixture_path}/movies/awesome'movie.mov")
        end

        it "should run ffmpeg successfully" do
          @movie.duration.should == 7.56
          @movie.frame_rate.should == 16.75
        end
      end

      context "given a non movie file" do
        before(:all) do
          @movie = MediaFile.new(__FILE__)
        end

        it "should not be valid" do
          @movie.should_not be_valid
        end

        it "should have a duration of 0" do
          @movie.duration.should == 0
        end

        it "should have nil height" do
          @movie.height.should be_nil
        end

        it "should have nil width" do
          @movie.width.should be_nil
        end

        it "should have nil frame_rate" do
          @movie.frame_rate.should be_nil
        end
      end

      context "given an empty flv file (could not find codec parameters)" do
        before(:all) do
          @movie = MediaFile.new("#{fixture_path}/movies/empty.flv")
        end

        it "should not be valid" do
          @movie.should_not be_valid
        end
      end

      context "given a broken mp4 file" do
        before(:all) do
          @movie = MediaFile.new("#{fixture_path}/movies/broken.mp4")
        end

        it "should not be valid" do
          @movie.should_not be_valid
        end

        it "should have nil calculated_aspect_ratio" do
          @movie.calculated_aspect_ratio.should be_nil
        end
      end

      context "given a weird aspect ratio file" do
        before(:all) do
          @movie = MediaFile.new("#{fixture_path}/movies/weird_aspect.small.mpg")
        end

        it "should parse the DAR" do
          @movie.dar.should == "704:405"
        end

        it "should have correct calculated_aspect_ratio" do
          @movie.calculated_aspect_ratio.to_s[0..14].should == "1.7382716049382" # substringed to be 1.9 compatible
        end
      end

      context "given an impossible DAR" do
        before(:all) do
          fake_output = StringIO.new(File.read("#{fixture_path}/outputs/file_with_weird_dar.txt"))
          Open3.stub(:popen3).and_yield(nil,nil,fake_output)
          @movie = MediaFile.new(__FILE__)
        end

        it "should parse the DAR" do
          @movie.dar.should == "0:1"
        end

        it "should calulate using width and height instead" do
          @movie.calculated_aspect_ratio.to_s[0..14].should == "1.7777777777777" # substringed to be 1.9 compatible
        end
      end

      context "given a weird storage/pixel aspect ratio file" do
        before(:all) do
          @movie = MediaFile.new("#{fixture_path}/movies/weird_aspect.small.mpg")
        end

        it "should parse the SAR" do
          @movie.sar.should == "64:45"
        end

        it "should have correct calculated_pixel_aspect_ratio" do
          @movie.calculated_pixel_aspect_ratio.to_s[0..14].should == "1.4222222222222" # substringed to be 1.9 compatible
        end
      end

      context "given a colorspace with parenthesis but no commas such as yuv420p(tv)" do
        before(:all) do
          fake_output = StringIO.new(File.read("#{fixture_path}/outputs/file_with_colorspace_with_parenthesis_but_no_comma.txt"))
          Open3.stub(:popen3).and_yield(nil,nil,fake_output)
          @movie = MediaFile.new(__FILE__)
        end

        it "should have correct video stream" do
          @movie.colorspace.should == "yuv420p(tv)"
        end
      end

      context "given an impossible SAR" do
        before(:all) do
          fake_output = StringIO.new(File.read("#{fixture_path}/outputs/file_with_weird_sar.txt"))
          Open3.stub(:popen3).and_yield(nil,nil,fake_output)
          @movie = MediaFile.new(__FILE__)
        end

        it "should parse the SAR" do
          @movie.sar.should == "0:1"
        end

        it "should using square SAR, 1.0 instead" do
          @movie.calculated_pixel_aspect_ratio.to_s[0..14].should == "1" # substringed to be 1.9 compatible
        end
      end

      context "given a file with ISO-8859-1 characters in output" do
        it "should not crash" do
          fake_output = StringIO.new(File.read("#{fixture_path}/outputs/file_with_iso-8859-1.txt"))
          Open3.stub(:popen3).and_yield(nil, nil, fake_output)
          expect { MediaFile.new(__FILE__) }.to_not raise_error
        end
      end

      context "given a file with 5.1 audio" do
        before(:all) do
          fake_output = StringIO.new(File.read("#{fixture_path}/outputs/file_with_surround_sound.txt"))
          Open3.stub(:popen3).and_yield(nil, nil, fake_output)
          @movie = MediaFile.new(__FILE__)
        end

        it "should have 6 audio channels" do
          @movie.audio_channels.should == 6
        end
      end

      context "given a file with no audio" do
        before(:all) do
          fake_output = StringIO.new(File.read("#{fixture_path}/outputs/file_with_no_audio.txt"))
          Open3.stub(:popen3).and_yield(nil, nil, fake_output)
          @movie = MediaFile.new(__FILE__)
        end

        it "should have nil audio channels" do
          @movie.audio_channels.should == nil
        end
      end

      context "given a file with non supported audio" do
        before(:all) do
          fake_output = StringIO.new(File.read("#{fixture_path}/outputs/file_with_non_supported_audio.txt"))
          Open3.stub(:popen3).and_yield(nil, nil, fake_output)
          @movie = MediaFile.new(__FILE__)
        end

        it "should not be valid" do
          @movie.should_not be_valid
        end
      end

      context "given a file with complex colorspace and decimal fps" do
        before(:all) do
          fake_output = StringIO.new(File.read("#{fixture_path}/outputs/file_with_complex_colorspace_and_decimal_fps.txt"))
          Open3.stub(:popen3).and_yield(nil, nil, fake_output)
          @movie = MediaFile.new(__FILE__)
        end

        it "should know the framerate" do
          @movie.frame_rate.should == 23.98
        end

        it "should know the colorspace" do
          @movie.colorspace.should == "yuv420p(tv, bt709)"
        end

        it "should know the width and height" do
          @movie.width.should == 960
          @movie.height.should == 540
        end
      end

      context "given an awesome movie file" do
        before(:all) do
          @movie = MediaFile.new("#{fixture_path}/movies/awesome movie.mov")
        end

        it "should remember the movie path" do
          @movie.path.should == "#{fixture_path}/movies/awesome movie.mov"
        end

        it "should parse duration to number of seconds" do
          @movie.duration.should == 7.56
        end

        it "should parse the bitrate" do
          @movie.bitrate.should == 481
        end

        it "should return nil rotation when no rotation exists" do
          @movie.rotation.should == nil
        end

        it "should parse the creation_time" do
          @movie.creation_time.should == Time.parse("2010-02-05 16:05:04")
        end

        it "should parse video stream information" do
          @movie.video_stream.should == "h264 (Main) (avc1 / 0x31637661), yuv420p(tv, bt709), 640x480 [SAR 1:1 DAR 4:3], 371 kb/s, 16.75 fps, 600 tbr, 600 tbn, 1200 tbc (default)"
        end

        it "should know the video codec" do
          @movie.video_codec.should =~ /h264/
        end

        it "should know the colorspace" do
          @movie.colorspace.should == "yuv420p(tv, bt709)"
        end

        it "should know the resolution" do
          @movie.resolution.should == "640x480"
        end

        it "should know the video bitrate" do
          @movie.video_bitrate.should == 371
        end

        it "should know the width and height" do
          @movie.width.should == 640
          @movie.height.should == 480
        end

        it "should know the framerate" do
          @movie.frame_rate.should == 16.75
        end

        it "should parse audio stream information" do
          @movie.audio_stream.should == "aac (mp4a / 0x6134706D), 44100 Hz, stereo, fltp, 75 kb/s (default)"
        end

        it "should know the audio codec" do
          @movie.audio_codec.should =~ /aac/
        end

        it "should know the sample rate" do
          @movie.audio_sample_rate.should == 44100
        end

        it "should know the number of audio channels" do
          @movie.audio_channels.should == 2
        end

        it "should know the audio bitrate" do
          @movie.audio_bitrate.should == 75
        end

        it "should should be valid" do
          @movie.should be_valid
        end

        it "should calculate the aspect ratio" do
          @movie.calculated_aspect_ratio.to_s[0..14].should == "1.3333333333333" # substringed to be 1.9 compatible
        end

        it "should know the file size" do
          @movie.size.should == 455546
        end

        it "should know the container" do
          @movie.container.should == "mov,mp4,m4a,3gp,3g2,mj2"
        end
      end
    end

    context "given a rotated movie file" do
      before(:all) do
        @movie = MediaFile.new("#{fixture_path}/movies/sideways movie.mov")
      end

      it "should parse the rotation" do
        @movie.rotation.should == 90
      end
    end

    describe "transcode" do
      it "should run the transcoder" do
        movie = MediaFile.new("#{fixture_path}/movies/awesome movie.mov")

        transcoder_double = double(Transcoder)
        Transcoder.should_receive(:new).
          with(movie, "#{tmp_path}/awesome.flv", {custom: "-vcodec libx264"}, preserve_aspect_ratio: :width).
          and_return(transcoder_double)
        transcoder_double.should_receive(:run)

        movie.transcode("#{tmp_path}/awesome.flv", {custom: "-vcodec libx264"}, preserve_aspect_ratio: :width)
      end
    end

    describe "screenshot" do
      it "should run the transcoder with screenshot option" do
        movie = MediaFile.new("#{fixture_path}/movies/awesome movie.mov")

        transcoder_double = double(Transcoder)
        Transcoder.should_receive(:new).
          with(movie, "#{tmp_path}/awesome.jpg", {seek_time: 2, dimensions: "640x480", screenshot: true}, preserve_aspect_ratio: :width).
          and_return(transcoder_double)
        transcoder_double.should_receive(:run)

        movie.screenshot("#{tmp_path}/awesome.jpg", {seek_time: 2, dimensions: "640x480"}, preserve_aspect_ratio: :width)
      end
    end
  end
end


lib/ffmpeg/movie.rb
require 'time'

module FFMPEG
  class Movie
    attr_reader :path, :duration, :time, :bitrate, :rotation, :creation_time
    attr_reader :video_stream, :video_codec, :video_bitrate, :colorspace, :resolution, :sar, :dar
    attr_reader :audio_stream, :audio_codec, :audio_bitrate, :audio_sample_rate
    attr_reader :container

    def initialize(path)
      raise Errno::ENOENT, "the file '#{path}' does not exist" unless File.exists?(path)

      @path = path

      # ffmpeg will output to stderr
      command = "#{FFMPEG.ffmpeg_binary} -i #{Shellwords.escape(path)}"
      output = Open3.popen3(command) { |stdin, stdout, stderr| stderr.read }

      fix_encoding(output)

      output[/Input \#\d+\,\s*(\S+),\s*from/]
      @container = $1

      output[/Duration: (\d{2}):(\d{2}):(\d{2}\.\d{2})/]
      @duration = ($1.to_i*60*60) + ($2.to_i*60) + $3.to_f

      output[/start: (\d*\.\d*)/]
      @time = $1 ? $1.to_f : 0.0

      output[/creation_time {1,}: {1,}(\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2})/]
      @creation_time = $1 ? Time.parse("#{$1}") : nil

      output[/bitrate: (\d*)/]
      @bitrate = $1 ? $1.to_i : nil

      output[/rotate\ {1,}:\ {1,}(\d*)/]
      @rotation = $1 ? $1.to_i : nil

      output[/Video:\ (.*)/]
      @video_stream = $1

      output[/Audio:\ (.*)/]
      @audio_stream = $1

      if video_stream
        commas_except_in_parenthesis = /(?:\([^()]*\)|[^,])+/ # regexp to handle "yuv420p(tv, bt709)" colorspace etc from http://goo.gl/6oi645
        @video_codec, @colorspace, resolution, video_bitrate = video_stream.scan(commas_except_in_parenthesis).map(&:strip)
        @video_bitrate = video_bitrate =~ %r(\A(\d+) kb/s\Z) ? $1.to_i : nil
        @resolution = resolution.split(" ").first rescue nil # get rid of [PAR 1:1 DAR 16:9]
        @sar = $1 if video_stream[/SAR (\d+:\d+)/]
        @dar = $1 if video_stream[/DAR (\d+:\d+)/]
      end

      if audio_stream
        @audio_codec, audio_sample_rate, @audio_channels, unused, audio_bitrate = audio_stream.split(/\s?,\s?/)
        @audio_bitrate = audio_bitrate =~ %r(\A(\d+) kb/s) ? $1.to_i : nil
        @audio_sample_rate = audio_sample_rate[/\d*/].to_i
      end

      @invalid = true if @video_stream.to_s.empty? && @audio_stream.to_s.empty?
      @invalid = true if output.include?("is not supported")
      @invalid = true if output.include?("could not find codec parameters")
    end

    def valid?
      not @invalid
    end

    def video?
      !@video_stream.to_s.empty?
    end

    def audio?
      !video? && !@audio_stream.to_s.empty?
    end

    def width
      resolution.split("x")[0].to_i rescue nil
    end

    def height
      resolution.split("x")[1].to_i rescue nil
    end

    def calculated_aspect_ratio
      aspect_from_dar || aspect_from_dimensions
    end

    def calculated_pixel_aspect_ratio
      aspect_from_sar || 1
    end

    def size
      File.size(@path)
    end

    def audio_channels
      return nil unless @audio_channels
      return @audio_channels[/\d*/].to_i if @audio_channels["channels"]
      return 1 if @audio_channels["mono"]
      return 2 if @audio_channels["stereo"]
      return 6 if @audio_channels["5.1"]
    end

    def frame_rate
      return nil unless video_stream
      video_stream[/(\d*\.?\d*)\s?fps/] ? $1.to_f : nil
    end

    def transcode(output_file, options = EncodingOptions.new, transcoder_options = {}, &block)
      Transcoder.new(self, output_file, options, transcoder_options).run &block
    end

    def screenshot(output_file, options = EncodingOptions.new, transcoder_options = {}, &block)
      Transcoder.new(self, output_file, options.merge(screenshot: true), transcoder_options).run &block
    end


    protected

    def aspect_from_dar
      return nil unless dar
      w, h = dar.split(":")
      aspect = w.to_f / h.to_f
      aspect.zero? ? nil : aspect
    end

    def aspect_from_sar
      return nil unless sar
      w, h = sar.split(":")
      aspect = w.to_f / h.to_f
      aspect.zero? ? nil : aspect
    end

    def aspect_from_dimensions
      aspect = width.to_f / height.to_f
      aspect.nan? ? nil : aspect
    end

    def fix_encoding(output)
      output[/test/] # Running a regexp on the string throws error if it's not UTF-8
    rescue ArgumentError
      output.force_encoding("ISO-8859-1")
    end

  end
end


lib/ffmpeg/io_ext.rb
require "timeout"
require "thread"

if RUBY_PLATFORM =~ /(win|w)(32|64)$/
  begin
    require 'win32/process'
  rescue LoadError
    "Warning: streamio-ffmpeg is missing the win32-process gem to properly handle hung transcodings. Install the gem (in Gemfile if using bundler) to avoid errors."
  end
end


# Monkey Patch timeout support into the IO class
class IO

  def each_with_timeout(pid, seconds, sep_string=$/)
    last_update = Time.now

    current_thread = Thread.current
    check_update_thread = Thread.new do
      loop do
        sleep 0.1
        if last_update - Time.now < -seconds
          current_thread.raise Timeout::Error.new("Output wait time expired.")
        end
      end
    end

    each(sep_string) do |buffer|
      last_update = Time.now
      yield buffer
    end
  rescue Timeout::Error
    if RUBY_PLATFORM =~ /(win|w)(32|64)$/
      Process.kill(1, pid)
    else
      Process.kill("SIGKILL", pid)
    end
    raise
  ensure
    check_update_thread.kill
  end

end




lib/streamio-ffmpeg.rb
$LOAD_PATH.unshift File.dirname(__FILE__)

require "logger"
require "stringio"

require "ffmpeg/version"
require "ffmpeg/errors"
require "ffmpeg/media_file"
require "ffmpeg/io_ext"
require "ffmpeg/transcoder"
require "ffmpeg/concatenator"
require "ffmpeg/encoding_options"


module FFMPEG

  def self.logger=(log)
    @logger = log
  end


  def self.logger
    return @logger if @logger
    logger = Logger.new(STDOUT)
    logger.level = Logger::INFO
    @logger = logger
  end


  def self.ffmpeg_binary=(bin)
    @ffmpeg_binary = bin
  end


  def self.ffmpeg_binary
    @ffmpeg_binary || "ffmpeg"
  end


  def self.concatenate(input_files, output_file, options = EncodingOptions.new, transcoder_options = {}, &block)
    raise unless input_files.respond_to? :each
    media_files = input_files.map do |f|
      f.is_a?(MediaFile) ? f : MediaFile.new(f)
    end
    Concatenator.new(media_files, output_file, options, transcoder_options).run &block
  end


  def self.test
    FFMPEG.concatenate([MediaFile.new("test/a.flac"),"test/b.wav","test/c.mp3"],"test/o.mp3") do |p|
      puts p
    end
  end

end




lib/ffmpeg/media_file.rb
require "time"


module FFMPEG
  class MediaFile
    attr_reader :path, :duration, :time, :bitrate, :rotation, :creation_time
    attr_reader :video_stream, :video_codec, :video_bitrate, :colorspace, :resolution, :sar, :dar
    attr_reader :audio_stream, :audio_codec, :audio_bitrate, :audio_sample_rate
    attr_reader :container


    def initialize(path)
      raise Errno::ENOENT, "The file '#{path}' does not exist." unless File.exists?(path)

      @path = path

      # ffmpeg will output to stderr
      command = "#{FFMPEG.ffmpeg_binary} -i #{Shellwords.escape(path)}"
      output = Open3.popen3(command) { |stdin, stdout, stderr| stderr.read }

      fix_encoding(output)

      output[/Input \#\d+\,\s*(\S+),\s*from/]
      @container = $1

      output[/Duration: (\d{2}):(\d{2}):(\d{2}\.\d{2})/]
      @duration = ($1.to_i*60*60) + ($2.to_i*60) + $3.to_f

      output[/start: (\d*\.\d*)/]
      @time = $1 ? $1.to_f : 0.0

      output[/creation_time {1,}: {1,}(\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2})/]
      @creation_time = $1 ? Time.parse("#{$1}") : nil

      output[/bitrate: (\d*)/]
      @bitrate = $1 ? $1.to_i : nil

      output[/rotate\ {1,}:\ {1,}(\d*)/]
      @rotation = $1 ? $1.to_i : nil

      output[/Video:\ (.*)/]
      @video_stream = $1

      output[/Audio:\ (.*)/]
      @audio_stream = $1

      if video_stream
        commas_except_in_parenthesis = /(?:\([^()]*\)|[^,])+/ # regexp to handle "yuv420p(tv, bt709)" colorspace etc from http://goo.gl/6oi645
        @video_codec, @colorspace, resolution, video_bitrate = video_stream.scan(commas_except_in_parenthesis).map(&:strip)
        @video_bitrate = video_bitrate =~ %r(\A(\d+) kb/s\Z) ? $1.to_i : nil
        @resolution = resolution.split(" ").first rescue nil # get rid of [PAR 1:1 DAR 16:9]
        @sar = $1 if video_stream[/SAR (\d+:\d+)/]
        @dar = $1 if video_stream[/DAR (\d+:\d+)/]
      end

      if audio_stream
        @audio_codec, audio_sample_rate, @audio_channels, unused, audio_bitrate = audio_stream.split(/\s?,\s?/)
        @audio_bitrate = audio_bitrate =~ %r(\A(\d+) kb/s) ? $1.to_i : nil
        @audio_sample_rate = audio_sample_rate[/\d*/].to_i
      end

      @invalid = true if @video_stream.to_s.empty? && @audio_stream.to_s.empty?
      @invalid = true if output.include?("is not supported")
      @invalid = true if output.include?("could not find codec parameters")
    end


    def valid?
      not @invalid
    end


    def video?
      !@video_stream.to_s.empty?
    end


    def audio?
      !video? && !@audio_stream.to_s.empty?
    end


    def width
      resolution.split("x")[0].to_i rescue nil
    end


    def height
      resolution.split("x")[1].to_i rescue nil
    end


    def calculated_aspect_ratio
      aspect_from_dar || aspect_from_dimensions
    end


    def calculated_pixel_aspect_ratio
      aspect_from_sar || 1
    end


    def size
      File.size(@path)
    end


    def audio_channels
      return nil unless @audio_channels
      return @audio_channels[/\d*/].to_i if @audio_channels["channels"]
      return 1 if @audio_channels["mono"]
      return 2 if @audio_channels["stereo"]
      return 6 if @audio_channels["5.1"]
    end


    def frame_rate
      return nil unless video_stream
      video_stream[/(\d*\.?\d*)\s?fps/] ? $1.to_f : nil
    end


    def transcode(output_file, options = EncodingOptions.new, transcoder_options = {}, &block)
      Transcoder.new(self, output_file, options, transcoder_options).run &block
    end


    def screenshot(output_file, options = EncodingOptions.new, transcoder_options = {}, &block)
      Transcoder.new(self, output_file, options.merge(screenshot: true), transcoder_options).run &block
    end


    protected

    def aspect_from_dar
      return nil unless dar
      w, h = dar.split(":")
      aspect = w.to_f / h.to_f
      aspect.zero? ? nil : aspect
    end


    def aspect_from_sar
      return nil unless sar
      w, h = sar.split(":")
      aspect = w.to_f / h.to_f
      aspect.zero? ? nil : aspect
    end


    def aspect_from_dimensions
      aspect = width.to_f / height.to_f
      aspect.nan? ? nil : aspect
    end


    def fix_encoding(output)
      output[/test/] # Running a regexp on the string throws error if it's not UTF-8
    rescue ArgumentError
      output.force_encoding("ISO-8859-1")
    end

  end

end


